P082: Feature/gps plugin & Serial port abstraction

TD-er opened the pull request:
New plugin:
GPS, for just about any GPS receiver using a serial port.
Still a number of features must be added.

Also made a wrapper around HW serial and SW serial and the HW serial interface for ESP32.

A plugin using a serial port now just has to call 2 functions, one for load and one for save to use this wrapper.
See _P082_GPS.ino as example.

The selector looks like this:

HW-serial port selected:
![image](https://user-images.githubusercontent.com/3751318/50388442-eab64700-0716-11e9-9567-0737684598ed.png)

Some software serial port selected:
![image](https://user-images.githubusercontent.com/3751318/50388450-0e798d00-0717-11e9-8da8-ff10fd7411c3.png)

N.B. when the combobox is set to some HW serial config, the RX/TX pin selectors will be hidden.

Still to do:
[ ] Add checkbox when RX or TX is not connected/needed (e.g. SDS011)
[ ] Add plugin using Serial1 (HW serial with TX only)
[ ] Some cleanup of the code to generate the GPIO names
[ ] Add connection statistics view to GPS plugin
[ ] Add some more checks to serial overflow
[ ] Add some more delays when reading/writing lots of data (using buffer maybe?)
[ ] Add some checks to disable serial port for log when HW serial0 is being used.
[ ] Let other code use the new `ESPeasySerial` wrapper class instead of `Serial` (checks for use of `swap()` )
[ ] Add some checks for use of `Serial1` on ESP32 (default pins are also used by flash => crash when using ESP32 Arduino defaults)
[ ] Check whether all other plugins using serial still work like expected.

TD-er committed:
[GPS] Add TinyGPSPlus library 1.0.2

[Source](https://github.com/mikalhart/TinyGPSPlus)

TD-er committed:
[GPS] Add TinyGPSPlus library 1.0.2

[Source](https://github.com/mikalhart/TinyGPSPlus)

TD-er committed:
[GPS] Improved precision

Applied PR from @qbazd
[lat, lng force math operations #47](https://github.com/mikalhart/TinyGPSPlus/pull/47)

TD-er committed:
[GPS] Improved precision

Applied PR from @qbazd
[lat, lng force math operations #47](https://github.com/mikalhart/TinyGPSPlus/pull/47)

TD-er committed:
[GPS] support for fix quality and fix mode

Applied this PR:
[support for fix quality and fix mode](https://github.com/mikalhart/TinyGPSPlus/pull/38)

TD-er committed:
[GPS] support for fix quality and fix mode

Applied this PR:
[support for fix quality and fix mode](https://github.com/mikalhart/TinyGPSPlus/pull/38)

TD-er committed:
[GPS] New plugin P082 GPS - initial commit

TD-er committed:
[GPS] New plugin P082 GPS - initial commit

TD-er committed:
Merge remote-tracking branch 'origin/feature/GPS_plugin' into feature/GPS_plugin

TD-er committed:
[GPS] Make helper object to abstract hardware/software serial

TD-er committed:
[GPS] Extend helper object for abstracting HW/SW serial

TD-er committed:
[GPS] Serial wrapper support including ESP32

TD-er committed:
[GPS] typo in JSON file caused build issues

TD-er committed:
[core 2.3.0] build attempt (failing due to iRAM)

Attempt to test core stability using core 2.3.0.
It compiles, but linking is not possible, since iram1_0_seg is full.

TD-er committed:
[Serial] Let plugins use new serial wrapper

TD-er committed:
[WiFi] Set to auto reconnect.

TD-er committed:
[Serial] Use uniform serial port selector for all serial plugins

TD-er committed:
[Stability] Add delay(0) to all successful calls to plugin functions

Since these calls can be performed in sequences, the time of these calls can become rather large.
Every call returning success did something, so add a delay(0) after such a plugin function call to make sure background tasks are performed.

TD-er commented:
Would be nice if some of you could already test it, since I will not be much around the computer tomorrow.
And I know there is some kind of feature-freeze at the moment, but I really needed the GPS plugin myself ;)
Also at this moment my mind is not really set to deep debugging, so this was some kind of distraction for me.

Grovkillen commented:
That is a nice Christmas present @TD-er good job. I'm waiting for my GPS units to arrive. Neo-6M, Neo-7M and Neo-M8n is incoming.

papperone commented:
wow that's great improvement!
Not sure if I can help testing but definitely I'll wait feedback on this before to rework the PZEM004T plugin as it relies only on serialcomm....

tedenda commented:
Very nice with a GPS Plugin!
I have a GPS module somewhere, will look for it and test it.

TD-er committed:
[build] Register ESPeasySerial as library in PlatformIO

TD-er commented:
I just updated the platformio config to be able to build it also when the library is not locally registered in PlatformIO.

The plugin is not yet finished, but the position and speed work and it should give an event when there is a fix.

TD-er committed:
[build] Added extra include of ESPeasySerial.h

Otherwise build depends on order of included files.

TD-er committed:
[#2197] Correct spelling error in 'Controller'

See #2197

TD-er committed:
[GPS] Allow GPS to update system time

TD-er committed:
[ESPEasy Serial] Fix build issues for core 2.4.0

TD-er committed:
[Puya] Backport core 2.5.0 Puya patch and make patched platform archives

This makes it possible to build Puya patched versions without having to patch the files first.

TD-er committed:
[JavaScript] Simple JS to update view in web interface on change serial

TD-er commented:
Would be nice if someone can also test this commit: https://github.com/letscontrolit/ESPEasy/pull/2194/commits/4744d638e4084ca4051df56ffc82f3f8c245e517
It allows anyone to build PUYA supported versions without having to patch the core libraries first.

TD-er committed:
[GPS] Add PPS pin as reference pilot for time sync

TD-er committed:
[GPS] Add documentation for P082_GPS.ino

TD-er committed:
[PUYA] Use newest core 2.5.0 backport patch on unpatched sources.

TD-er committed:
[GPS] Include the GPS docs in the Spinx tree.
